modulus/distributed/utils.py

```diff
@@ -218,3 +219,293 @@ def _gather(input_, dim_, group=None):
+        output = output.to(dtype=tensor.dtype)
+    else:
+        # else: just do sum in native dtype
+        output = tmp.sum(dim=stack_dim)
+
+    return output
+
+
+def gather_v_wrapper(
+    tensor: torch.Tensor,
+    sizes: List[int],
+    dim: int = 0,
+    dst: int = 0,
+    group: Optional[dist.ProcessGroup] = None,
+) -> torch.Tensor:
+    """implementation of a distributed gather_v primitive"""
+
+    comm_size = dist.get_world_size(group=group)
+    rank = dist.get_rank(group=group)
+    assert len(sizes) == comm_size
+    assert 0 <= dst < comm_size
+    assert dim < tensor.dim()
+
+    global_size = sum(sizes)
+
+    if comm_size == 1:
+        return tensor
+
+    tensor_shape = list(tensor.shape)
+    tensor_list = [None] * comm_size
+
+    if rank == dst:
+        for r in range(comm_size):
+            req_list = [None] * comm_size
+            if r == dst:
+                tensor_list[dst] = tensor
+            else:
+                tensor_shape[dim] = sizes[r]
+                tensor_list[r] = torch.empty(
+                    tensor_shape,
+                    device=tensor.device,
+                    dtype=tensor.dtype,
+                )
+
+                req_list[r] = dist.irecv(tensor_list[r], src=r, group=group)
+
+        for r in range(comm_size):
+            if r != dst:
+                req_list[r].wait()
+
+        output = torch.cat(tensor_list, dim=dim)
+
+    else:
+        req = dist.isend(tensor, dst=dst, group=group)
+        req.wait()
+
+        tensor_shape[dim] = sum(sizes)
+        output = torch.empty(
+            tensor_shape,
+            device=tensor.device,
+            dtype=tensor.dtype,
+        )
+
+    return output
+
+
+def scatter_v_wrapper(
+    tensor: torch.Tensor,
+    sizes: List[int],
+    dim: int = 0,
+    src: int = 0,
+    group: Optional[dist.ProcessGroup] = None,
+) -> torch.Tensor:
+    """implementation of a distributed scatter_v primitive"""
+
+    comm_size = dist.get_world_size(group=group)
+    rank = dist.get_rank(group=group)
+    assert len(sizes) == comm_size
+    assert 0 <= src < comm_size
+    assert dim < tensor.dim()
+
+    global_size = sum(sizes)
+    local_size = sizes[rank]
+
+    assert global_size == tensor.size(dim)
+
+    tensor_shape = list(tensor.shape)
+    tensor_shape[dim] = sizes[rank]
+    output = torch.empty(
+        tensor_shape,
+        dtype=tensor.dtype,
+        device=tensor.device,
+    )
+
+    # dist.scatter doesn't support tensors of different shape
+    # so this implementation is using explicit send/recv combinations
+    scatter_list = None
+    if rank == src:
+        scatter_list = torch.split(tensor, sizes, dim=dim)
+        req_list = [None] * comm_size
+        for r in range(comm_size):
+            tensor_to_scatter_to_r = scatter_list[r]
+            if r == src:
+                output = tensor_to_scatter_to_r
+            else:
+                req_list[r] = dist.isend(tensor_to_scatter_to_r, dst=r, group=group)
```

Comment:
Same thing with scatter: https://pytorch.org/docs/stable/distributed.html#torch.distributed.scatter

Reply:
IIUC, scatter definitely says `Note that all Tensors in scatter_list must have the same size.`